rfcs/014-pagination.md

```diff
@@ -0,0 +1,181 @@
+---
+name: Pagination support
+about: Add bookmark based pagination interface
+title: 'Pagination API'
+labels: rfc, discussion
+assignees: ''
+
+---
+
+Implement pagination support for query like endpoints
+
+# Introduction
+
+The main driver for this proposal is the introduction of FoundationDB as a
+storage engine. FoundationDB imposes limits on transaction duration and size.
+Therefore we need to find way to restrict amount of data we return to customers.
+We could simply set the maximum `limit` which would cap amount of rows client can
+request. However this "solution" has a big disadvantage. Which is it would require
+clients to write pagination recepy in their code. Current pagination scheme
+requires complex logic on the client side. There are quite a few corner cases
+to handle.
+
+## Abstract
+
+The main addition is to add new bookmark based pagination scheme to all quiery like endpoints.
+As a first step we would keep _all_dbs and _changes out of scope for the following reasons:
+- _all_dbs - the endpoint returns a list instead of object as the rest of the endpoints
+- _changes - the endpoint contains too many different modes and it would require more careful consideration
+
+The endpoints in scope are:
+- {db}/_all_docs
+- {db}/_all_docs/queries
+- _dbs_info
+- {db}/_design/{ddoc}/_view/{view}
+- {db}/_design/{ddoc}/_view/{view}/queries
+
+In a nutshell the idea is:
+- add `page_size` query field to control number of rows on each page and to flag
+  that client expects paginated response
+- add `first`, `previous`, `next` fields which contain bookmark part of URI
+- add `bookmark` query field to retrieve bookmarked page
+
+
+## Requirements Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this
+document are to be interpreted as described in
+[RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.txt).
+
+## Terminology
+
+- `bookmark` - is opaque token which would contain information needed to retrieve
+  bookmarked page. The format of the token value MUST NOT be relied upon in the client.
+
+---
+
+# Detailed Description
+
+## Implementation proposal
+
+1) Add new optional query field called `bookmark` to following endpoints
+  - {db}/_all_docs
+  - {db}/_all_docs/queries
+  - _dbs_info
+  - {db}/_design/{ddoc}/_view/{view}
+  - {db}/_design/{ddoc}/_view/{view}/queries
+
+2) Add new optional query field called `page_size` if it is set we would use
+  paginated endpoint otherwise use old code path
+3) Implement per-endpoint configurable max limits to restrict page size of
+  paginated responses
+
+ ```
+ [request_limits]
+ _all_docs = 5000
+ _all_docs/queries = 5000
+ _all_dbs = 5000
+ _dbs_info = 5000
+ _view = 2500
+ _view/queries = 2500
+ _find = 2500
+ ```
+4) Add following additional fields into response:
+  ```
+    "first": "/myddb/_all_docs?bookmark=12345678945621321689",
```

Comment:
I think it would be nicer if these were objects that included _both_ a standalone bookmark and the link.
e.g.
```json
"first": {
  "bookmark": "12345678945621321689",
  "href": "/myddb/_all_docs?bookmark=12345678945621321689"
}
```
For browser based usage I think the link/`href` is useful, but for parameter based code the DB and endpoint are likely configured separately from the `bookmark` e.g. functions like:
`get_all_docs(db='mydb', bookmark='12345678945621321689')`

Providing only a link/`href` form of the bookmark would require that type of client-side code to either parse and extract the `bookmark` token from the link or to implement additional link-based versions of functions in addition to parameter-based.

Supplying the bookmark in both standalone and link form satisfies both cases.

Reply:
I am thinking about removing href version and was about to send a message about it to ML. The problem with href is doc id size. Currently the max doc id size is configurable. In production we see doc id sizes as big as 7K. FDB imposes a limit on key path. Therefore we would probably have a strict limit in CouchDB 4.0. Let's assume the limit would be set to 4K. The bookmark would encode all parameters passed in the first query. This could easily exceed the max URL length supported by the browser. In such case we would have to use POST request. Which means href wouldn't make sense.

I didn't send the message to ML yet because I am trying to find the doc id limit in previous discussions.